File: recursion/python/solutions/last.py
"""Solution for last."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend


def last(lst):
    """
    Return the last element of a non-empty list.

    last(1 -> 2 -> 3) => 3
    last(42) => 42
    """
    if is_empty(lst.next):
        return lst.value

    first, rest = unprepend(lst)

    return last(rest)


if __name__ == '__main__':
    # Test cases
    assert last(Node.from_list([1, 2, 3])) == 3
    assert last(Node.from_list([42])) == 42
    assert last(Node.from_list([1, 2, 3, 4, 5])) == 5
    assert last(Node.from_list(['a', 'b', 'c'])) == 'c'

    print("All tests passed!")

File: recursion/javascript/solutions/last.js
/**
 * Solution for last.
 */

const { Node, isEmpty, unprepend } = require('./linkedList');

function last(list) {
  if (isEmpty(list.next)) {
    return list.value;
  }

  let [first, rest] = unprepend(list);

  return last(rest);
}

if (require.main === module) {
  console.assert(last(Node.fromArray([1, 2, 3])) === 3);
  console.assert(last(Node.fromArray([42])) === 42);
  console.assert(last(Node.fromArray([1, 2, 3, 4, 5])) === 5);

  console.log('All tests passed!');
}

module.exports = { last };
